rust-src/concordium_base/src/web3id/mod.rs

```diff
@@ -615,9 +615,101 @@ impl<C: Curve, AttributeType: Attribute<C::Scalar>> crate::common::Serial
 pub enum Web3IdChallengeMarker {}
 
-/// Challenge string that serves as a distinguishing context when requesting
+/// Sha256 challenge string that serves as a distinguishing context when requesting
+/// proofs.
+pub type Sha256Challenge = HashBytes<Web3IdChallengeMarker>;
+
+/// Context challenge that serves as a distinguishing context when requesting
 /// proofs.
-pub type Challenge = HashBytes<Web3IdChallengeMarker>;
+#[derive(
+    Clone,
+    Eq,
+    PartialEq,
+    Ord,
+    PartialOrd,
+    serde::Deserialize,
+    serde::Serialize,
+    crate::common::Serial,
+    Debug,
+)]
+pub struct ContextChallenge {
+    /// This part of the challenge is suppose to be provided by the dapp backend (e.g. merchant backend).
+    pub given: Vec<GivenContext>,
+    /// This part of the challenge is suppose to be provided by the wallet or ID app.
+    pub requested: Vec<GivenContext>,
+}
+
+#[derive(
+    Clone,
+    Eq,
+    PartialEq,
+    Ord,
+    PartialOrd,
+    serde::Deserialize,
+    serde::Serialize,
+    crate::common::Serial,
+    Debug,
+)]
+pub struct GivenContext {
+    pub label: String,
+    pub context: ContextValue,
+}
+
+#[derive(Clone, Eq, PartialEq, Ord, PartialOrd, serde::Deserialize, serde::Serialize, Debug)]
+#[serde(rename_all = "camelCase")]
+pub enum ContextValue {
+    Sha256(Sha256Challenge),
+    BlockHash(BlockHash),
+    String(String),
+}
+
+impl crate::common::Serial for ContextValue {
+    fn serial<B: crate::common::Buffer>(&self, out: &mut B) {
+        match self {
+            ContextValue::Sha256(sha256) => {
+                // Serialize the tag first.
+                0u8.serial(out);
+                sha256.serial(out)
+            }
+            ContextValue::BlockHash(block_hash) => {
+                // Serialize the tag first.
+                1u8.serial(out);
+                block_hash.serial(out)
+            }
+            ContextValue::String(string) => {
+                // Serialize the tag first.
+                2u8.serial(out);
+                string.serial(out)
+            }
+        }
+    }
+}
+
+/// A challenge that can be added to the proof transcript.
+#[derive(serde::Deserialize, serde::Serialize)]
+// The type is `untagged` to be backward compatible with old proofs and requests.
+#[serde(untagged)]
+#[derive(Clone, Eq, PartialEq, Ord, PartialOrd, Debug)]
+pub enum Challenge {
+    Sha256(Sha256Challenge),
+    V1(ContextChallenge),
+}
+
+impl crate::common::Serial for Challenge {
+    fn serial<B: crate::common::Buffer>(&self, out: &mut B) {
+        match self {
+            Challenge::Sha256(hash_bytes) => {
+                // No tag is added to be backward compatible with old proofs and requests.
+                hash_bytes.serial(out)
+            }
+            Challenge::V1(context_info) => {
+                // Serialize the version/tag first.
+                1u8.serial(out);
+                context_info.serial(out)
+            }
+        }
+    }
+}
```

Comment:
This serialization doesn't seem like it's injective. Is it actually needed? Is it ever deserialized?